man/man1/map.1
.TH MAP 1
.SH NAME
map, mapdemo, mapd \- draw maps on various projections
.SH SYNOPSIS
.B map
.I projection
[
.I option ...
]
.PP
.B mapdemo
.PP
.SH DESCRIPTION
.I Map
prepares on the standard output a
map suitable for display by any
plotting filter described in
.IR  plot (1).
A menu of projections is produced in response to an unknown
.IR projection .
.I Mapdemo
is a short course in mapping.
.PP
The default data for
.I map
are world shorelines.
Option
.B -f
accesses more detailed data
classified by feature.
.TP
.BR -f " [ \fIfeature\fR ... ]"
Features are ranked 1 (default) to 4 from major to minor.
Higher-numbered ranks include all lower-numbered ones.
Features are
.RS
.TF country[1-3]
.TP
.BR shore [ 1 - 4 ] 
seacoasts, lakes, and islands; option
.B -f
always shows
.B shore1
.TP
.BR ilake [ 1 - 2 ] 
intermittent lakes
.TP
.BR river [ 1 - 4 ] 
rivers
.TP
.BR iriver [ 1 - 3 ] 
intermittent rivers
.TP
.BR canal [ 1 - 3 ] 
.BR 3 =irrigation
canals
.TP
.BR glacier
.TP
.BR iceshelf [ 12 ] 
.TP
.BR reef
.TP
.BR saltpan [ 12 ] 
.TP
.BR country [ 1 - 3 ] 
.BR 2 =disputed
boundaries,
.BR 3 =indefinite
boundaries
.TP
.BR state
states and provinces (US and Canada only)
.PD
.RE
.PP
In other options
coordinates are in degrees, with north latitude
and west longitude counted as positive.
.TP 0
.BI -l " S N E W"
Set the southern and northern latitude
and the eastern and western longitude limits.
Missing arguments are filled out from the list
\-90, 90, \-180, 180,
or lesser limits suitable to the
projection at hand.
.TP
.BI -k " S N E W
Set the scale as if for a map with limits
.B -l
.I "S N E W"\f1.
Do not consider any
.B -l
or
.B -w
option in setting scale.
.TP
.BI -o " lat lon rot"
Orient the map in a nonstandard position.
Imagine a transparent gridded sphere around the globe.
Turn the overlay about the North Pole
so that the Prime Meridian (longitude 0)
of the overlay coincides with meridian
.I lon
on the globe.
Then tilt the North Pole of the
overlay along its Prime Meridian to latitude
.I lat
on the globe.
Finally again turn the
overlay about its `North Pole' so
that its Prime Meridian coincides with the previous position
of meridian
.IR rot .
Project the map in
the standard form appropriate to the overlay, but presenting
information from the underlying globe.
Missing arguments are filled out from the list
90, 0, 0.
In the absence of
.BR - o ,
the orientation is 90, 0,
.IR m ,
where
.I m
is the middle of the longitude range.
.TP
.BI -w " S N E W"
Window the map by the specified latitudes
and longitudes in the tilted, rotated coordinate system.
Missing arguments are filled out from the list \-90, 90, \-180, 180.
(It is wise to give an encompassing
.B -l
option with
.BR -w .
Otherwise for small windows computing time
varies inversely with area!)
.TP
.BI -d " n"
For speed, plot only every
.IR n th
point.
.TP
.B  -r
Reverse left and right
(good for star charts and inside-out views).
.ns
.TP
.B -v
Verso.
Switch to a normally suppressed sheet of the map, such as the
back side of the earth in orthographic projection.
.TP
.B  -s1
.br
.ns
.TP
.B -s2
Superpose; outputs for a
.B -s1
map (no closing) and a
.B -s2
map (no opening) may be concatenated.
.TP
.BI -g " dlat dlon res"
Grid spacings are
.IR dlat ,
.IR dlon .
Zero spacing means no grid.
Missing
.I dlat
is taken to be zero.
Missing
.I dlon
is taken the same as
.IR dlat .
Grid lines are drawn to a resolution of
.I res
(2° or less by default).
In the absence of
.BR - g ,
grid spacing is 10°.
.TP
.BI -p " lat lon extent"
Position the point
.I lat, lon
at the center of the plotting area.
Scale the map so that the height (and width) of the
nominal plotting area is
.I extent
times the size of one degree of latitude
at the center.
By default maps are scaled and positioned
to fit within the plotting area.
An
.I extent
overrides option
.BR -k .
.TP
.BI -c " x y rot"
After all other positioning and scaling operations
have been performed, rotate the image
.I rot
degrees counterclockwise about the center 
and move the center to position
.IR x ,
.IR y ,
where the nominal plotting area is
.RI \-1≤ x ≤1,
.RI \-1≤ y ≤1.
Missing arguments are taken to be 0.
.BR -x
Allow the map to extend outside the nominal plotting area.
.TP
.BR -m " [ \fIfile\fP ... ]"
Use
map data from named files.
If no files are named, omit map data.
Names that do not exist as pathnames are looked up in
a standard directory, which contains, in addition to the
data for
.BR -f ,
.RS
.LP
.TF counties
.TP
.B world
World Data Bank I (default)
.TP
.B states
US map from Census Bureau
.TP
.B counties
US map from Census Bureau
.PD
.RE
.IP
The environment variables
.B MAP 
and
.B MAPDIR 
change the default
map and default directory.
.TP
.BI -b " \fR[\fPlat0 lon0 lat1 lon1\fR... ]"
Suppress the drawing of the normal boundary
(defined by options
.BR -l 
and
.BR -w ).
Coordinates, if present, define the vertices of a
polygon to which the map is clipped.
If only two vertices are given, they are taken to be the
diagonal of a rectangle.
To draw the polygon, give its vertices as a
.B -u
track.
.TP
.BI -t " file ..."
The
.I files
contain lists of points,
given as latitude-longitude pairs in degrees.
If the first file is named 
.LR - ,
the standard input is taken instead.
The points of each list are plotted as connected `tracks'.
.IP
Points in a track file may be followed by label strings.
A label breaks the track.
A label may be prefixed by
\fL"\fR,
.LR : ,
or 
.L !
and is terminated by a newline.
An unprefixed string or a string prefixed with
.L
"
is displayed at the designated point.
The first word of a
.L :
or
.L !
string names a special symbol (see option
.BR -y ).
An optional numerical second word is a scale factor
for the size of the symbol, 1 by default.
A
.L :
symbol is aligned with its top to the north; a
.L !
symbol is aligned vertically on the page.
.TP
.BI -u " file ..."
Same as
.BR -t ,
except the tracks are
unbroken lines.
.RB ( -t
tracks appear as dot-dashed lines if the plotting filter supports them.)
.TP
.BI -y " file
The
.I file
contains 
.MR plot 7 -style
data for
.L :
or
.L !
labels in
.B -t
or
.B -u
files.
Each symbol is defined by a comment
.BI : name
then a sequence of
.L m
and
.L v
commands.
Coordinates (0,0) fall on the plotting point.
Default scaling is as if the nominal plotting range were
.LR "ra -1 -1 1 1" ;
.L ra
commands in
.I file
change the scaling.
.SS Projections
Equatorial projections centered on the Prime Meridian
(longitude 0).
Parallels are straight horizontal lines.
.PP
.PD 0
.TP 1.5i
.B mercator
equally spaced straight meridians, conformal,
straight compass courses
.TP
.B sinusoidal
equally spaced parallels,
equal-area, same as
.LR "bonne 0" .
.TP
.BI cylequalarea " lat0"
equally spaced straight meridians, equal-area,
true scale on
.I lat0
.TP
.B cylindrical
central projection on tangent cylinder
.TP
.BI rectangular " lat0"
equally spaced parallels, equally spaced straight meridians, true scale on
.I lat0
.TP
.BI gall " lat0"
parallels spaced stereographically on prime meridian, equally spaced straight
meridians, true scale on
.I lat0
.TP
.B mollweide
(homalographic) equal-area, hemisphere is a circle
.br
.B gilbert()
sphere conformally mapped on hemisphere and viewed orthographically
.TP
.B gilbert
globe mapped conformally on hemisphere, viewed orthographically
.PD
.PP
Azimuthal projections centered on the North Pole.
Parallels are concentric circles.
Meridians are equally spaced radial lines.
.PP
.PD 0
.TP 1.5i
.B azequidistant
equally spaced parallels,
true distances from pole
.TP
.B azequalarea
equal-area
.TP
.B gnomonic
central projection on tangent plane,
straight great circles
.TP
.BI perspective " dist"
viewed along earth's axis
.I dist
earth radii from center of earth
.TP
.B orthographic
viewed from infinity
.TP
.B stereographic
conformal, projected from opposite pole
.TP
.B laue
.IR radius " = tan(2\(mu" colatitude ),
used in X-ray crystallography
.TP
.BI fisheye " n"
stereographic seen from just inside medium with refractive index
.I n
.TP
.BI newyorker " r"
.IR radius " = log(" colatitude / r ):
.I New Yorker
map from viewing pedestal of radius
.I r
degrees
.PD
.PP
Polar conic projections symmetric about the Prime Meridian.
Parallels are segments of concentric circles.
Except in the Bonne projection,
meridians are equally spaced radial
lines orthogonal to the parallels.
.PP
.PD 0
.TP 1.5i
.BI conic " lat0"
central projection on cone tangent at
.I lat0
.TP
.BI simpleconic " lat0 lat1"
equally spaced parallels, true scale on
.I lat0
and
.I lat1
.TP
.BI lambert " lat0 lat1"
conformal, true scale on 
.I lat0
and 
.I lat1
.TP
.BI albers " lat0 lat1"
equal-area, true scale on
.I lat0
and 
.I lat1
.TP
.BI bonne " lat0"
equally spaced parallels, equal-area,
parallel
.I lat0
developed from tangent cone
.PD
.PP
Projections with bilateral symmetry about
the Prime Meridian
and the equator.
.PP
.PD 0
.TP 1.5i
.B polyconic
parallels developed from tangent cones,
equally spaced along Prime Meridian
.TP
.B aitoff
equal-area projection of globe onto 2-to-1
ellipse, based on 
.I azequalarea
.TP
.B lagrange
conformal, maps whole sphere into a circle
.TP
.BI bicentric " lon0"
points plotted at true azimuth from two
centers on the equator at longitudes
.IR ±lon0 ,
great circles are straight lines
(a stretched
.IR gnomonic
)
.TP
.BI elliptic " lon0"
points plotted at true distance from
two centers on the equator at longitudes
.I ±lon0
.TP
.B globular
hemisphere is circle,
circular arc meridians equally spaced on equator,
circular arc parallels equally spaced on 0- and 90-degree meridians
.TP
.B vandergrinten
sphere is circle,
meridians as in
.IR globular ,
circular arc parallels resemble 
.I mercator
.PD
.PP
Doubly periodic conformal projections.
.PP
.TP 1.5i
.B guyou
W and E hemispheres are square
.PD 0
.TP
.B square
world is square with Poles
at diagonally opposite corners
.TP
.B tetra
map on tetrahedron with edge
tangent to Prime Meridian at S Pole,
unfolded into equilateral triangle
.TP
.B hex
world is hexagon centered
on N Pole, N and S hemispheres are equilateral
triangles
.PD
.PP
Miscellaneous projections.
.PP
.PD 0
.TP 1.5i
.BI harrison " dist angle"
oblique perspective from above the North Pole,
.I dist
earth radii from center of earth, looking
along the Date Line
.I angle
degrees off vertical
.TP
.BI trapezoidal " lat0 lat1"
equally spaced parallels,
straight meridians equally spaced along parallels,
true scale at
.I lat0
and
.I lat1
on Prime Meridian
.PD
.br
.B lune(lat,angle)
conformal, polar cap above latitude
.I lat
maps to convex lune with given
.I angle
at 90\(deE and 90\(deW
.PP
Retroazimuthal projections.
At every point the angle between vertical and a straight line to
`Mecca', latitude
.I lat0
on the prime meridian,
is the true bearing of Mecca.
.PP
.PD 0
.TP 1.5i
.BI mecca " lat0"
equally spaced vertical meridians
.TP
.BI homing " lat0"
distances to Mecca are true
.PD
.PP
Maps based on the spheroid.
Of geodetic quality, these projections do not make sense
for tilted orientations.
For descriptions, see corresponding maps above.
.PP
.PD 0
.TP 1.5i
.B sp_mercator
.TP
.BI sp_albers " lat0 lat1"
.SH EXAMPLES
.TP
.L
map perspective 1.025 -o 40.75 74
A view looking down on New York from 100 miles
(0.025 of the 4000-mile earth radius) up.
The job can be done faster by limiting the map so as not to `plot'
the invisible part of the world:
.LR "map perspective 1.025 -o 40.75 74 -l 20 60 30 100".
A circular border can be forced by adding option
.LR "-w 77.33" .
(Latitude 77.33° falls just inside a polar cap of
opening angle arccos(1/1.025) = 12.6804°.)
.TP
.L
map mercator -o 49.25 -106 180
An `equatorial' map of the earth
centered on New York.
The pole of the map is placed 90\(de away (40.75+49.25=90)
on the
other side of the earth.
A 180° twist around the pole of the map arranges that the
`Prime Meridian' of the map runs from the pole of the
map over the North Pole to New York
instead of down the back side of the earth.
The same effect can be had from
.L
map mercator -o 130.75 74
.TP
.L
map albers 28 45 -l 20 50 60 130 -m states
A customary curved-latitude map of the United States.
.TP
.L
map harrison 2 30 -l -90 90 120 240 -o 90 0 0
A fan view covering 60° on either
side of the Date Line, as seen from one earth radius
above the North Pole gazing at the
earth's limb, which is 30° off vertical.
The
.B -o
option overrides the default
.BR "-o 90 0 180" , 
which would rotate
the scene to behind the observer.
.SH FILES
.TF /lib/map/[1-4]??
.TP
.B /lib/map/[1-4]??
World Data Bank II, for
.B -f
.TP
.B /lib/map/*
maps for
.B -m
.TP
.B /lib/map/*.x
map indexes
.TP
.B mapd
Map driver program
.SH SOURCE
.B \*9/src/cmd/map
.SH "SEE ALSO"
.IR map (7), 
.MR plot 1
.SH DIAGNOSTICS
`Map seems to be empty'\(ema coarse survey found
zero extent within the 
.B -l
and
.BR -w 
bounds; for maps of limited extent
the grid resolution,
.IR res ,
or the limits may have to be refined.
.SH BUGS
Windows (option
.BR -w )
cannot cross the Date Line.
No borders appear along edges arising from
visibility limits.
Segments that cross a border are dropped, not clipped.
Excessively large scale or
.B -d
setting may cause long line segments to be dropped.
.I Map
tries to draw grid lines dotted and
.B -t
tracks dot-dashed.
As very few plotting filters properly support
curved textured lines, these lines are likely to
appear solid.
The west-longitude-positive convention
betrays Yankee chauvinism.
.I Gilbert
should be a map from sphere to sphere, independent of
the mapping from sphere to plane.


CHANGES
# Note: the name listed next to a change sometimes
# submitted the patch but more often just reported
# the problem.

# For entries more recent than January 30, 2008,
# see the Git repository history at 
# https://github.com/9fans/plan9port/.

January 30, 2008
	acme: fix regexp match choice bug
	devdraw: add F11 for full-screen toggle
	rio: accomodate full-screen apps more gracefully
	sam: fix regexp match choice bug

January 28, 2008
	xd: fix printing of 4-byte values on 64-bit machines (Sam Hopkins)

January 12, 2008
	delatex: reset lexer state at each beginning of file
	ipso: do not flush factotum keys unless using factotum file
	libdiskfs: fix ext2 directory reading bug (Mike Mammarella)
	lp: move devices to devices.sample
	rio: make full-screen work properly; add showevent
	venti: fewer calls to gettimeofday
	vmount: need to say -t nfs now on Linux

January 10, 2008
	acme: make STACK bigger for OS X (David Swasey)
	dist,unix: use MKSHELL=$PLAN9/bin/rc (John DeGood)
	libdraw: fix spinning event loop (David Bulkow)
	libregexp: fix match choice bug

December 7, 2007
	acme, sam, libregexp: regexp fixes and reversions

December 4, 2007
	hoc: print exact floating point output
	sam: turn off DEBUG logging (oops)
	vbackup: fix uninitialized variable use in -i code

November 28, 2007
	acme: refresh window properly after Tab command (Fazlul Shahriar)

November 27, 2007
	.hgignore: ignore more
	libdiskfs: make ext2 use little-endian disk format on big-endian machines (Joshua Wood)
	acme: accept clicks between body and scroll bar (Axel Belinfante)
	tpic: clean up better
	upas/smtp: remove intermediate files
	vbackup: add -i flag to avoid score prefetching (David Swasey)
	venti: add debugging HTML form

November 25, 2007
	acme: avoid infinite loop on search in 8125-byte file (Roger Peppe)
	lib9: avoid DIOCGMEDIASIZE on FreeBSD 4.11 (Rodrigo Readi)
	mailfs: use root in imapcopylist (Richard Bilson)
	moveplan9.files: add tmac/tmac.an (Tony Lainson)

November 23, 2007
	9pfuse: ignore Linux O_CLOEXEC flag (Michael Kaminsky)

November 22, 2007
	9pfuse: fix array bounds error (Michael Teichgräber)
	vbackup: fix fsysopen error message (Joshua Wood)

November 7, 2007
	9fs: use -n to mount sources (Jeff Sickel),
	9pfuse: MacFUSE moved again! (Jeff Sickel)

November 6, 2007
	.cvsignore: add 9pm (for old trees)
	unvf: ignore case in Content-Transfer-Encoding: line.
	validateattachment: reject malformed zip files

November 5, 2007
	building: use ar s flag to avoid ranlib (OS X)
	diff: continue after i/o errors when diffing many files
	factotum: silence gcc warning
	getns: turn / into _ in $DISPLAY
	libdiskfs: OS X linker fixes
	libmach: OS X linker fixes
	libmp: no text jump labels (OS X)
	libthread: use our own swapcontext on OS X
	mailfs: handle NIL body, add -r flag (Richard Bilson)

November 4, 2007
	man: document hg interface, fix CVS host name

November 3, 2007
	libthread: more OS X 10.5 changes (Jeff Sickel)

October 31, 2007
	venti: use 9ar, not ar (fixes Darwin build)

October 30, 2007
	libthread: OS X 10.5 tweaks (Jeff Sickel, Bakul Shah)

October 29, 2007
	grap: fclose, error context bug fixes (Doug Gwyn)
	venti: fix sync deadlock, add /proc stub

October 12, 2007
	9p: add -n to disable authentication
	9pserve: add -n to reject authentication
	srv: add -n to reject authentication

October 9, 2007
	mount: ask for Unix test explicitly (Chris King)

September 24-25, 2007
	venti: many fixes, new index cache

September 16, 2007
	factotum: fix build error (Axel Belinfante)

September 15, 2007
	auxstats: use acpi for battery on Linux (Andrey Mirtchovski)

September 13, 2007
	factotum: implement 9pcr (of dubious utility)

August 28, 2007
	9pserve: fix infinite loop bug

August 23, 2007
	venti: add -s flag, bug fixes (Michael Kaminsky)

August 22, 2007
	9.rc: change to work with rc -e (Tony Lainson)
	9pserve: fix races causing assert(c->nmsg == 0) to fail.
	acme: fix warning memory leak (Michael Teichgräber)
	libfmt: add fmtlocale to Makefile (Faried Nawaz)
	libthread: simplify alt tracking, possibly fixing bug
	mailfs: fix handling of FLAGS responses (Robert Vollmert)
	unix ports: Linux x86-64 Makefiles (Ævar Arnfjörð Bjarmason)

July 25, 2007
	9pfuse: even more MacFUSE path changes

June 29, 2007
	dict: look in $dictpath for dictionaries

June 24, 2007
	mailfs: fix subject line in header file.

June 23, 2007
	mailfs: various IMAP bug fixes
	Mail: add Get command, refresh at startup,
		fix deleted messages bug
	nedmail: refresh at startup, on i and y.

June 20, 2007
	9pfuse: add -a aname to get at sourcesdump

June 19, 2007
	9pfuse: the MacFUSE file locations changed again (Rob Pike)

June 18, 2007
	upas/marshal: remove debugging print
	9pserve: ignore window resizes (fix really should be elsewhere)
	libdiskfs: more HFS fixes (David Swasey)

June 15, 2007
	acme(1): fix Snarf bug, change Indent ON to apply to all windows.

June 12, 2007
	vcat: always write last block, to make disk correct size
	vbackup: more OS X fixes, add -M flag
	vbackup(8): document -M

June 9, 2007
	libregexp, sam, acme: day 1 bug in sam regexp code

June 8, 2007
	libdiskfs: add HFS resource fork support (David Swasey)

June 5, 2007
	ramfs: be less chatty

June 3, 2007
	vbackup: fix dotdot (David Swasey)

May 29, 2007
	libdiskfs: more HFS fixes (David Swasey)
	
May 28, 2007
	libdiskfs: missing Bterm, make ext2 quieter,
		add disk partitions, HFS code (David Swasey)

May 21, 2007
	libregexp: fix pointer comparison warning (Ramon de Vera)
	test: fix
	
May 20, 2007
	touch: print error if wstat fails
	libc: use futimes on Linux to implement dirfwstat

May 19, 2007
	test: import from Plan 9

May 18, 2007
	libregexp: fix bug about leading ^ in regexp on multiline text (Roger Peppe)

May 15, 2007
	libc: new media size code for _p9dir on FreeBSD (Lou Kamenov)
	acme(4): document tab width field in ctl file

May 10, 2007
	libc: more NaN fixes (rsc)
		use device names in _p9dir for FreeBSD (Lou Kamenov)

src/libmach/map.c
/*
 * File map routines
 */
#include <u.h>
#include <libc.h>
#include <bio.h>
#include <mach.h>

static int fdrw(Map*, Seg*, u64int, void*, uint, int);
static int zerorw(Map*, Seg*, u64int, void*, uint, int);
static int mrw(Map*, u64int, void*, uint, int);
static int datarw(Map*, Seg*, u64int, void*, uint, int);

Map*
allocmap(void)
{
	return mallocz(sizeof(Map), 1);
}

void
freemap(Map *map)
{
	if(map == nil)
		return;
	free(map->seg);
	free(map);
}

int
addseg(Map *map, Seg seg)
{
	Seg *ss;

	if(map == nil){
		werrstr("invalid map");
		return -1;
	}

	ss = realloc(map->seg, (map->nseg+1)*sizeof(ss[0]));
	if(ss == nil)
		return -1;
	map->seg = ss;
	if(seg.rw == 0){
		if(seg.name && strcmp(seg.name, "zero") == 0)
			seg.rw = zerorw;
		else if(seg.p)
			seg.rw = datarw;
		else
			seg.rw = fdrw;
	}
	map->seg[map->nseg] = seg;
	return map->nseg++;
}

int
findseg(Map *map, char *name, char *file)
{
	int i;

	if(map == nil)
		return -1;
	for(i=0; i<map->nseg; i++){
		if(name && (!map->seg[i].name || strcmp(map->seg[i].name, name) != 0))
			continue;
		if(file && (!map->seg[i].file || strcmp(map->seg[i].file, file) != 0))
			continue;
		return i;
	}
	werrstr("segment %s in %s not found", name, file);
	return -1;
}

int
addrtoseg(Map *map, u64int addr, Seg *sp)
{
	int i;
	Seg *s;

	if(map == nil){
		werrstr("no map");
		return -1;
	}
	for(i=map->nseg-1; i>=0; i--){
		s = &map->seg[i];
		if(s->base <= addr && addr-s->base < s->size){
			if(sp)
				*sp = *s;
			return i;
		}
	}
	werrstr("address 0x%lux is not mapped", addr);
	return -1;
}

int
addrtosegafter(Map *map, u64int addr, Seg *sp)
{
	int i;
	Seg *s, *best;
	ulong bdist;

	if(map == nil){
		werrstr("no map");
		return -1;
	}

	/*
	 * If segments were sorted this would be easier,
	 * but since segments may overlap, sorting also
	 * requires splitting and rejoining, and that's just
	 * too complicated.
	 */
	best = nil;
	bdist = 0;
	for(i=map->nseg-1; i>=0; i--){
		s = &map->seg[i];
		if(s->base > addr){
			if(best==nil || s->base-addr < bdist){
				bdist = s->base - addr;
				best = s;
			}
		}
	}
	if(best){
		if(sp)
			*sp = *best;
		return best-map->seg;
	}
	werrstr("nothing mapped after address 0x%lux", addr);
	return -1;
}

void
removeseg(Map *map, int i)
{
	if(map == nil)
		return;
	if(i < 0 || i >= map->nseg)
		return;
	memmove(&map->seg[i], &map->seg[i+1], (map->nseg-(i+1))*sizeof(Seg));
	map->nseg--;
}

int
get1(Map *map, u64int addr, uchar *a, uint n)
{
	return mrw(map, addr, a, n, 1);
}

int
get2(Map *map, u64int addr, u16int *u)
{
	u16int v;

	if(mrw(map, addr, &v, 2, 1) < 0)
		return -1;
	*u = mach->swap2(v);
	return 2;
}

int
get4(Map *map, u64int addr, u32int *u)
{
	u32int v;

	if(mrw(map, addr, &v, 4, 1) < 0)
		return -1;
	*u = mach->swap4(v);
	return 4;
}

int
get8(Map *map, u64int addr, u64int *u)
{
	u64int v;

	if(mrw(map, addr, &v, 8, 1) < 0)
		return -1;
	*u = mach->swap8(v);
	return 8;
}

int
geta(Map *map, u64int addr, u64int *u)
{
	u32int v;

	if(machcpu == &machamd64)
		return get8(map, addr, u);
	if(get4(map, addr, &v) < 0)
		return -1;
	*u = v;
	return 4;
}

int
put1(Map *map, u64int addr, uchar *a, uint n)
{
	return mrw(map, addr, a, n, 0);
}

int
put2(Map *map, u64int addr, u16int u)
{
	u = mach->swap2(u);
	return mrw(map, addr, &u, 2, 0);
}

int
put4(Map *map, u64int addr, u32int u)
{
	u = mach->swap4(u);
	return mrw(map, addr, &u, 4, 0);
}

int
put8(Map *map, u64int addr, u64int u)
{
	u = mach->swap8(u);
	return mrw(map, addr, &u, 8, 0);
}

static Seg*
reloc(Map *map, u64int addr, uint n, u64int *off, uint *nn)
{
	int i;
	ulong o;

	if(map == nil){
		werrstr("invalid map");
		return nil;
	}

	for(i=map->nseg-1; i>=0; i--){
		if(map->seg[i].base <= addr){
			o = addr - map->seg[i].base;
			if(o >= map->seg[i].size)
				continue;
			if(o+n > map->seg[i].size)
				*nn = map->seg[i].size - o;
			else
				*nn = n;
			*off = o;
			return &map->seg[i];
		}
	}
	werrstr("address 0x%lux not mapped", addr);
	return nil;
}

static int
mrw(Map *map, u64int addr, void *a, uint n, int r)
{
	uint nn;
	uint tot;
	Seg *s;
	u64int off;

	for(tot=0; tot<n; tot+=nn){
		s = reloc(map, addr+tot, n-tot, &off, &nn);
		if(s == nil)
			return -1;
		if(s->rw(map, s, off, a, nn, r) < 0)
			return -1;
	}
	return 0;
}

static int
fdrw(Map *map, Seg *seg, u64int addr, void *a, uint n, int r)
{
	int nn;
	uint tot;
	ulong off;

	USED(map);
	off = seg->offset + addr;
	for(tot=0; tot<n; tot+=nn){
		if(r)
			nn = pread(seg->fd, a, n-tot, off+tot);
		else
			nn = pwrite(seg->fd, a, n-tot, off+tot);
		if(nn < 0)
			return -1;
		if(nn == 0){
			werrstr("partial %s at address 0x%lux in %s",
				r ? "read" : "write", off+tot, seg->file);
			return -1;
		}
	}
	return 0;
}

static int
zerorw(Map *map, Seg *seg, u64int addr, void *a, uint n, int r)
{
	USED(map);
	USED(seg);
	USED(addr);

	if(r==0){
		werrstr("cannot write zero segment");
		return -1;
	}
	memset(a, 0, n);
	return 0;
}

static int
datarw(Map *map, Seg *seg, u64int addr, void *a, uint n, int r)
{
	USED(map);

	if(r)
		memmove(a, seg->p+addr, n);
	else
		memmove(seg->p+addr, a, n);
	return 0;
}
